### PR TITLE
Issue 16795 improve savepublish 2

### PR DESCRIPTION
**Pull request by jdotcms**

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/util/ThreadContextUtil.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import com.dotmarketing.util.UtilMethods;
+
+/**
+ * Thread Context Util provides methods to handle reindex stuff and other thread local context things
+ * @author jsancas
+ */
+public class ThreadContextUtil {
+
+    private static ThreadLocal<ThreadContext> contextLocal = new ThreadLocal<>();
+    private static final ThreadContextUtil     INSTANCE    = new ThreadContextUtil();
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/util/ThreadContextUtil.java]: ThreadContextUtil should be a static class - no reason to use a singleton pattern here.

File: dotCMS/src/main/java/com/dotcms/util/ThreadContextUtil.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    public ThreadContext getContext () {
+
+        return contextLocal.get();
+    }
+
+
+    /**
+     * Set the current thread context
+     */
+    public void setContext (final ThreadContext  context) {
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/util/ThreadContextUtil.java]: why have a setContext? ThreadContext is mutable (which is fine)

File: dotCMS/src/main/java/com/dotcms/util/ThreadContextUtil.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+    public static ThreadContextUtil getInstance () {
+
+        return INSTANCE;
+    }
+    /**
+     * Get the context from the current thread
+     * @return {@link ThreadContext}
+     */
+    public ThreadContext getContext () {
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/util/ThreadContextUtil.java]: should this be `getOrCreateContext()`?

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java [hunk 4]
```diff
@@ -2694,10 +2697,17 @@ public Contentlet fireContentWorkflow(final Contentlet contentlet, final Content
 		this.validateActionStepAndWorkflow(contentlet, dependencies.getModUser());
 		this.checkShorties (contentlet);
 
-		final WorkflowProcessor processor = this.fireWorkflowPreCheckin(contentlet, dependencies.getModUser());
+		final ThreadContextUtil contextUtil = ThreadContextUtil.getInstance();
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java]: Just make this a call to a static method, I think

File: dotCMS/src/main/java/com/dotcms/util/ThreadContextUtil.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+    /**
+     * Wrap a void method into not reindex call
+     * @param delegate {@link VoidDelegate}
+     * @throws Exception
+     */
+    public void wrapVoidNoReindex (final VoidDelegate delegate) {
+
+        final ThreadContext threadContext =
+                UtilMethods.get(INSTANCE.getContext(), ()->new ThreadContext());
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/util/ThreadContextUtil.java]: create a method called `getOrCreateContext()` and use that

**Review by wezell — CHANGES_REQUESTED** [review]
Much cleaner than the contentlet params, though I did request changes.  Also, we need some tests.

File: dotCMS/src/main/java/com/dotcms/util/ThreadContextUtil.java (hunk 1 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/util/ThreadContextUtil.java]: Done

File: dotCMS/src/main/java/com/dotcms/util/ThreadContextUtil.java (hunk 3 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/util/ThreadContextUtil.java]: Done

File: dotCMS/src/main/java/com/dotcms/util/ThreadContextUtil.java (hunk 2 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/util/ThreadContextUtil.java]: Done

File: dotCMS/src/main/java/com/dotcms/util/ThreadContextUtil.java (hunk 5 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/util/ThreadContextUtil.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java (hunk 4 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java]: Done

**Review by jgambarios — APPROVED** [review]

**Review by wezell — APPROVED** [review]

**Pull request merged**